src/main/java/techreborn/blockentity/machine/tier1/GreenhouseControllerBlockEntity.java

```diff
@@ -0,0 +1,236 @@
+					|| block instanceof CactusBlock
+			) {
+				if (canUseEnergy(TechRebornConfig.greenhouseControllerEnergyPerBonemeal)) {
+					useEnergy(TechRebornConfig.greenhouseControllerEnergyPerBonemeal);
+					blockState.scheduledTick((ServerWorld) world, blockPos, world.random);
+				}
+			}
+		}
+		
+		if (block instanceof CropBlock) {
+			// Normal crop
+			CropBlock crop = (CropBlock) block;
+			if (crop.isMature(blockState)) {
+				if (canUseEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest)) {
+					useEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest);
+					insertIntoInv(Block.getDroppedStacks(blockState, (ServerWorld) world, blockPos, null));
+					world.setBlockState(blockPos, crop.withAge(0), 2);
+				}
+			}
+		} else if (block instanceof NetherWartBlock) {
+			// Why it isn't a CropBlock? :(
+			if (blockState.get(NetherWartBlock.AGE) >= 3) {
+				if (canUseEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest)) {
+					useEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest);
+					insertIntoInv(Block.getDroppedStacks(blockState, (ServerWorld) world, blockPos, null));
+					world.setBlockState(blockPos, blockState.with(NetherWartBlock.AGE, 0), 2);
+				}
+			}
+		} else if (block instanceof SweetBerryBushBlock) {
+			if (blockState.get(SweetBerryBushBlock.AGE) >= 3) {
+				if (canUseEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest)) {
+					useEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest);
+					insertIntoInv(Block.getDroppedStacks(blockState, (ServerWorld) world, blockPos, null));
+					world.setBlockState(blockPos, blockState.with(SweetBerryBushBlock.AGE, 1), 2);
+				}
+			}
+		} else if (block instanceof CocoaBlock) {
+			if (blockState.get(CocoaBlock.AGE) >= 2) {
+				if (canUseEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest)) {
+					useEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest);
+					insertIntoInv(Block.getDroppedStacks(blockState, (ServerWorld) world, blockPos, null));
+					world.setBlockState(blockPos, blockState.with(CocoaBlock.AGE, 0), 2);
+				}
+			}
+		} else if (block instanceof GourdBlock) {
+			// Pumpkin, Melon
+			if (canUseEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest)) {
+				useEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest);
+				insertIntoInv(Block.getDroppedStacks(blockState, (ServerWorld) world, blockPos, null));
+				world.breakBlock(blockPos, false);
+			}
+		} else if (block instanceof SugarCaneBlock
+				|| block instanceof CactusBlock
+				|| block instanceof BambooBlock
+		) {
+			for (int y = 1; world.getBlockState(blockPos.up(y)).getBlock() == block; y++) {
+				if (canUseEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest)) {
+					useEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest);
+					insertIntoInv(Block.getDroppedStacks(blockState, (ServerWorld) world, blockPos.up(y), null));
+				}
+				world.breakBlock(blockPos.up(y), false);
+			}
+		}
+		
+	}
```

Comment:
There is a fair amount of duplicated code here, 

```java
if (canUseEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest)) {
	useEnergy(TechRebornConfig.greenhouseControllerEnergyPerHarvest);
	insertIntoInv(Block.getDroppedStacks(blockState, (ServerWorld) world, blockPos, null));
	world.setBlockState(blockPos, blockState.with(SweetBerryBushBlock.AGE, 1), 2);
}
```

is mostly the same across all of them. I would make a basic api out of it to slove the duplicate code and the isuse of the voiding items.

Something like this could work well.

```java
import net.minecraft.block.BlockState;
import net.minecraft.item.ItemStack;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

import java.util.List;

interface CropHandler {

	/**
	 * Used to break a crop block and return its drops. Returns a list of item stacks, can be empty but never null.
	 * removeFromWorld can be used to check if the inv has space before removing from the world
	 * 
	 * @param world The {@link World}
	 * @param blockPos The {@link BlockPos} of the crop, may not be the block to break (sugar cane 1 up)
	 * @param blockState the {@link BlockState} of the block
	 * @param removeFromWorld when true the block is removed from the world, when false it still returns the items to allow checking for output slots.
	 * @return A {@link List} of {@link ItemStack} to insert into the inventory
	 */
	List<ItemStack> tryBreak(World world, BlockPos blockPos, BlockState blockState, boolean removeFromWorld);
}
```

Reply:
LogBlock harvest missing. 
Vanilla saplings are instance of Fertilizable. Thus I assume that vanilla trees are supported.